pkg/controller/elasticsearch/driver/bootstrap.go

```diff
@@ -23,6 +25,18 @@ func AnnotatedForBootstrap(cluster v1alpha1.Elasticsearch) bool {
 
 func ReconcileClusterUUID(c k8s.Client, cluster *v1alpha1.Elasticsearch, observedState observer.State) error {
 	if AnnotatedForBootstrap(*cluster) {
```

Comment:
I think there's a possible race condition here:

- initial state: one 6.x bootstrapped master node
- loop n:
  - user request an upgrade to 7.2
  - `elasticsearch.k8s.elastic.co/cluster-uuid` is removed
  - spec is changed .... but K8S is lagging and the new master is not rolled out immediately 
- loop n+1:
  - 6.8 master is still there, uid is set again
  - 7.2 finally comes out, `uid` is set: no bootstrap 💥 

I guess a fix should be to always  check if the current master is out of sync with something like that:

```golang
	if clusterIsBootstrapped(observedState) && !reBootstrap {
		// cluster bootstrapped but not annotated yet
		return annotateWithUUID(cluster, observedState, c)
	}
```

_nit: An even more subtile race condition would be to have an upscale request right after `elasticsearch.k8s.elastic.co/cluster-uuid` is removed. All new master Pods would be added simultaneously, but I guess we are fine because Zen2 would be bootstrapping anyway._

Reply:
Well spotted!

> An even more subtile race condition would be to have an upscale request right after elasticsearch.k8s.elastic.co/cluster-uuid is removed. All new master Pods would be added simultaneously, but I guess we are fine because Zen2 would be bootstrapping anyway.

Yes that would be fine because as soon as we have at least one other master still running when we roll out the 7.2 node it will bootstrap OK.